imitation/util/buffer.py

```diff
@@ -0,0 +1,213 @@
+
+
+class Buffer:
+  """A buffer that stores numpy arrays of fixed shape, and from which these
+  arrays can be arbitrarily sampled."""
+
+  capacity: int
+  """The number of data samples that can be stored in this buffer."""
+
+  sample_shape: Tuple[int]
+  """The shape of each data sample stored in this buffer."""
+
+  _buffer: np.ndarray
+  """The underlying numpy array (which actually stores the data)."""
+
+  _n_data: int
+  """An integer in `range(0, self.capacity + 1)`. The number of
+  samples currently stored in this buffer. This attribute is the return
+  value of `self.__len__`."""
+
+  _idx: int
+  """An integer in `range(0, self.capacity)`.
+  The index of the first row that new data should be written to."""
+
+  def __init__(self, capacity: int, sample_shape: Tuple[int, ...], dtype):
+    """Constructs a Buffer.
+
+    Args:
+        capacity: The number of samples that can be stored.
+        sample_shape: The shape of each sample stored in this buffer.
+        dtype (`np.dtype`-like): The numpy dtype that will be stored.
+    """
+    self.capacity = capacity
+    self.sample_shape = tuple(sample_shape)
+    self._buffer = np.zeros((capacity,) + self.sample_shape, dtype=dtype)
+    self._n_data = 0
+    self._idx = 0
+
+  @classmethod
+  def fromData(cls, data: np.ndarray) -> "Buffer":
+    """Construct and return a Buffer containing only the provided data.
+
+    The returned ReplayBuffer is at full capacity and ready for sampling.
+    """
+    capacity, *sample_shape = data.shape
+    buf = cls(capacity, sample_shape, dtype=data.dtype)
+    buf.store(data)
+    return buf
+
+  def store(self, data: np.ndarray) -> None:
+    """Store new data samples in the buffer, replacing old samples with FIFO
+    priority.
+
+    Args:
+        data: An array with shape `(n_samples,) + self.sample_shape`, where
+            `n_samples` is less than or equal to `self.capacity`.
+    Raises:
+        ValueError: `data` is empty.
+        ValueError: If `n_samples` is greater than `self.capacity`.
+        ValueError: data is the wrong shape.
+    """
+    if len(data) == 0:
+      raise ValueError("Trying to store empty data.")
+    if len(data) > self.capacity:
+      raise ValueError("Not enough capacity to store data.")
+    if data.shape[1:] != self.sample_shape:
+      raise ValueError("Wrong data_shape")
+
+    new_idx = self._idx + len(data)
+    if new_idx > self.capacity:
+      # import pdb; pdb.set_trace()
+      n_remain = self.capacity - self._idx
+      # Need to loop around the buffer. Break into two recursive calls.
+      self.store(data[:n_remain])
+      assert self._idx == 0
+      self.store(data[n_remain:])
+    else:
+      self._buffer[self._idx:new_idx] = data
+      self._idx = (self._idx + len(data)) % self.capacity
+      self._n_data = min(self._n_data + len(data), self.capacity)
+
+  def sample(self, n_samples: int) -> np.ndarray:
+    """Uniformly sample `n_samples` samples from the buffer without replacement.
```

Comment:
The sampling is actually *with* replacement. Which is probably fine.

Reply:
Yes, this should be with replacement which was intended